source/beamline/BioXAS/BioXASSSRLMonochromatorRegionControl.cpp

```diff
@@ -0,0 +1,846 @@
+#include "BioXASSSRLMonochromatorRegionControl.h"
+#include "BioXASSSRLMonochromator.h"
+
+BioXASSSRLMonochromatorRegionControl::BioXASSSRLMonochromatorRegionControl(QObject *parent) :
+	AMCompositeControl("RegionControl", "", parent, "BioXAS SSRL Monochromator Region Control")
+{
+	// Initialize local variables.
+
+	value_ = BioXASSSRLMonochromator::Region::None;
+	setpoint_ = BioXASSSRLMonochromator::Region::None;
+	moveInProgress_ = false;
+
+	upperSlit_ = 0;
+	lowerSlit_ = 0;
+	slitsStatus_ = 0;
+	paddle_ = 0;
+	paddleStatus_ = 0;
+	keyStatus_ = 0;
+	brakeStatus_ = 0;
+	bragg_ = 0;
+	braggAtCrystalChangePositionStatus_ = 0;
+	crystalChange_ = 0;
+	crystalChangeCWLimitStatus_ = 0;
+	crystalChangeCCWLimitStatus_ = 0;
+	regionAStatus_ = 0;
+	regionBStatus_ = 0;
+
+	// Initialize inherited variables.
+
+	setEnumStates(QStringList() << "A" << "B" << "None");
+	setMoveEnumStates(QStringList() << "A" << "B");
+	setAllowsMovesWhileMoving(false);
+	setContextKnownDescription("Region Control");
+
+	// Make connections
+
+	connect( this, SIGNAL(connected(bool)), this, SIGNAL(enumChanged()) );
+	connect( this, SIGNAL(valueChanged(double)), this, SIGNAL(enumChanged()) );
+	connect( this, SIGNAL(setpointChanged(double)), this, SIGNAL(enumChanged()) );
+}
+
+BioXASSSRLMonochromatorRegionControl::~BioXASSSRLMonochromatorRegionControl()
+{
+
+}
+
+double BioXASSSRLMonochromatorRegionControl::minimumValue() const
+{
+	return BioXASSSRLMonochromator::Region::None;
+}
+
+double BioXASSSRLMonochromatorRegionControl::maximumValue() const
+{
+	return BioXASSSRLMonochromator::Region::B;
+}
+
+bool BioXASSSRLMonochromatorRegionControl::canMove() const
+{
+	bool exists = false;
+	bool connected = false;
+	bool canMove = false;
+
+	exists = (
+		upperSlit_ &&
+		lowerSlit_ &&
+		slitsStatus_ &&
+		paddle_ &&
+		paddleStatus_ &&
+		keyStatus_ &&
+		brakeStatus_ &&
+		bragg_ &&
+		braggAtCrystalChangePositionStatus_ &&
+		crystalChange_ &&
+		crystalChangeCWLimitStatus_ &&
+		crystalChangeCCWLimitStatus_ &&
+		regionAStatus_ &&
+		regionBStatus_
+	);
+
+	if (exists) {
+		connected = isConnected();
+
+		canMove = (
+			!moveInProgress_ &&
+
+			upperSlit_->canMove() &&
+			lowerSlit_->canMove() &&
+			paddle_->canMove() &&
+			bragg_->canMove() &&
+			crystalChange_->canMove()
+		);
+	}
+
+	return (exists && connected && canMove);
+}
+
+AMControl::FailureExplanation BioXASSSRLMonochromatorRegionControl::move(double setpoint)
+{
+	if (!isConnected())
+		return AMControl::NotConnectedFailure;
+
+	if (isMoving())
+		return AMControl::AlreadyMovingFailure;
+
+	if (!canMove())
+		return AMControl::LimitFailure;
+
+	// Now we can assume that the region control is ready for a crystal change.
+	// Update the saved setpoint value.
+
+	setSetpoint((int)setpoint);
+
+	if (!validRegionSetpoint(setpoint_)) {
+		return AMControl::LimitFailure;
```

Comment:
Would this be a good place to put an AMErrorMon of some kind?  Maybe even with some of it's state checks in the code above it?